File: frontend/src/freshness.ts
import type { Snapshot } from './types'

export function evidenceExpired(snapshot: Snapshot | null | undefined, now = Date.now()): boolean {
  return !!snapshot?.options.some(option => {
    const ttl = option.kind === 'hotel' ? 300000 : option.kind === 'activity' ? 30 * 86400000 : null
    return ttl !== null && [...option.costs.map(cost => cost.evidence), ...(option.supporting_evidence || [])].some(evidence => {
      const age = now - Date.parse(evidence.retrieved_at)
      return !Number.isFinite(age) || age < 0 || age > ttl
    })
  })
}

File: backend/src/travel_agent/freshness.py
from datetime import datetime, timezone


def evidence_expired(snapshot) -> bool:
    if snapshot is None:
        return False
    now = datetime.now(timezone.utc)
    for option in snapshot.options:
        ttl = 300 if option.kind == 'hotel' else 30 * 86400 if option.kind == 'activity' else None
        if ttl is not None:
            for item in [*(cost.evidence for cost in option.costs), *option.supporting_evidence]:
                age = (now - item.retrieved_at).total_seconds()
                if age < 0 or age > ttl:
                    return True
    return False
